### PR TITLE
Changes as an effect of changing persistent storage model.

### PR DESCRIPTION
**Pull request by rohe**
Instead of having class instances doing the storage we instead want to put the burden on the frontend.
This will among other things decrease the number of DB calls significantly.
The frontend will have the methods dump, flush and load to store, clear and retrieve state.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -22,7 +22,7 @@ exclude_lines = [
 
 [tool.poetry]
 name = "cryptojwt"
-version = "1.4.1"
+version = "1.4.2"
```
Comment (jschlyter) [inline pyproject.toml]: This seems like a minor version bump, 1.5.0 perhaps?

File: src/cryptojwt/key_bundle.py [hunk 2]
```diff
@@ -751,7 +751,7 @@ def difference(self, bundle):
 
         return [k for k in self._keys if k not in bundle]
 
-    def dump(self):
+    def dump(self, cutoff: Optional[list] = None):
```
Comment (jschlyter) [inline src/cryptojwt/key_bundle.py]: Somewhat confusing attribute name (at least for me). Perhaps `excluded_attrs` is better?

The type should be `Optional[List[str]]` I think (I assume attribute names are strings)

File: src/cryptojwt/key_bundle.py [hunk 3]
```diff
@@ -1246,3 +1280,19 @@ def init_key(filename, type, kid="", **kwargs):
     _new_key = key_gen(type, kid=kid, **kwargs)
     dump_jwk(filename, _new_key)
     return _new_key
+
+
+def key_by_alg(alg: str):
+    if alg.startswith("RS"):
+        return key_gen("RSA", alg="RS256")
```
Comment (jschlyter) [inline src/cryptojwt/key_bundle.py:1295]: `alg` is not used by `key_gen` I believe, remove?

File: src/cryptojwt/key_issuer.py [hunk 4]
```diff
@@ -350,16 +351,18 @@ def __len__(self):
             nr += len(kb)
         return nr
 
-    def dump(self, exclude=None):
+    def dump(self, exclude=None, cutoff: Optional[list] = None) -> dict:
         """
         Returns the content as a dictionary.
 
+        :param exclude: Issuer that should not be include in the dump
```
Comment (jschlyter) [inline src/cryptojwt/key_issuer.py]: - `exclude_issuer` (perhaps make it `exclude_issuers` and a `Optional[List[str]]` to allow multiple issuers to be excluded?)
- `exclude_attrs` (or `exclude_attributes`)

File: src/cryptojwt/key_jar.py [hunk 5]
```diff
@@ -32,7 +32,6 @@ def __init__(
         remove_after=3600,
         httpc=None,
         httpc_params=None,
-        storage=None,
```
Comment (jschlyter) [inline src/cryptojwt/key_jar.py:35]: Can we assume that no one (excluding closing family) has used the storage class yet?

**Review by jschlyter — CHANGES_REQUESTED** [review]

File: pyproject.toml (hunk 1 above)
Comment (rohe) [reply pyproject.toml]: OK

File: src/cryptojwt/key_bundle.py (hunk 3 above)
Comment (rohe) [reply src/cryptojwt/key_bundle.py:1295]: Actually it is possible to use.
For the case where you want to limit the usage of for instance an RSA key to only one algorithm.

File: src/cryptojwt/key_bundle.py (hunk 2 above)
Comment (rohe) [reply src/cryptojwt/key_bundle.py]: Horticultural term :-)
Imaging that you are dealing with something that looks like a tree and there are branches you don't want to be included so you cut them off. 
Yes, attribute names are strings.

File: src/cryptojwt/key_jar.py (hunk 5 above)
Comment (rohe) [reply src/cryptojwt/key_jar.py:35]: Yes !

File: src/cryptojwt/key_issuer.py (hunk 4 above)
Comment (rohe) [reply src/cryptojwt/key_issuer.py]: OK then.

File: src/cryptojwt/key_bundle.py [hunk 6]
```diff
@@ -751,7 +752,7 @@ def difference(self, bundle):
 
         return [k for k in self._keys if k not in bundle]
 
-    def dump(self):
+    def dump(self, exclude_attribute: Optional[List[str]] = None):
```
Comment (jschlyter) [inline src/cryptojwt/key_bundle.py]: `exclude_attributes`

File: src/cryptojwt/key_issuer.py [hunk 7]
```diff
@@ -350,16 +352,17 @@ def __len__(self):
             nr += len(kb)
         return nr
 
-    def dump(self, exclude=None):
+    def dump(self, exclude_attribute: Optional[List[str]] = None) -> dict:
```
Comment (jschlyter) [inline src/cryptojwt/key_issuer.py]: `exclude_attributes`

File: src/cryptojwt/key_jar.py [hunk 8]
```diff
@@ -635,44 +626,79 @@ def copy(self):
...
-    def dump(self, exclude=None):
+    def dump(
+        self,
+        exclude_issuer: Optional[List[str]] = None,
+        exclude_attribute: Optional[List[str]] = None,
+    ) -> dict:
         """
         Returns the key jar content as dictionary
 
+        :param exclude_issuer: A list of issuers you don't want included.
```
Comment (jschlyter) [inline src/cryptojwt/key_jar.py]: `exclude_issuers`?

**Review by jschlyter — DISMISSED** [review]
All good, perhaps we should use plural for some kwargs (see comments)

File: src/cryptojwt/key_bundle.py (hunk 6 above)
Comment (rohe) [reply src/cryptojwt/key_bundle.py]: OK

File: src/cryptojwt/key_issuer.py (hunk 7 above)
Comment (rohe) [reply src/cryptojwt/key_issuer.py]: OK

File: src/cryptojwt/key_jar.py (hunk 8 above)
Comment (rohe) [reply src/cryptojwt/key_jar.py]: OK

**Review by jschlyter — APPROVED** [review]

**Pull request merged**